rpc-broker/src/policy.c

```diff
@@ -0,0 +1,290 @@
+
+    char *ruleptr;
+    const char *delimiter = " ";
+    char *token = strtok_r(rule, delimiter, &ruleptr);
+
+    if (!token)
+        return -1;
+
+    current->policy = token[0] == 'a';
+    token = strtok_r(NULL, delimiter, &ruleptr);
+
+    while (token) {
+
+        char *field = strtok_r(NULL, delimiter, &ruleptr);
+
+        if (!field && token[0] != 's') {
+            return -1;
+        } else if (!strcmp("destination", token)) {
+            current->destination = strdup(field);
+        } else if (!strcmp("dom-type", token)) {
+            current->domtype = strdup(field);
+        } else if (!strcmp("interface", token)) {
+            current->interface = strdup(field);
+        } else if (!strcmp("if-boolean", token)) {
+            current->if_bool = strdup(field);
+            token = strtok_r(NULL, delimiter, &ruleptr);
+            current->if_bool_flag = strcmp("true", token) == 0 ? 1 : 0;
+        } else if (!strcmp("stubdom", token)) {
+            current->stubdom = 1;
+        } else if (!strcmp("path", token)) {
+            current->path = strdup(field);
+        } else if (!strcmp("member", token)) {
+            current->member = strdup(field);
+        } else if (!strcmp("out-any", token)) {
+            current->out = 1;
+        } else {
+            DBUS_BROKER_WARNING("Unrecognized Rule-Token: %s", token);
+            free_rule(*current);
+            current = NULL;
+            return -1;
+        }
+
+        token = strtok_r(NULL, delimiter, &ruleptr);
+    }
+
+    return 0;
+}
+
+static inline void get_rules(DBusConnection *conn, struct domain_policy *dom)
+{
+    int rule_idx;
+
+    for (rule_idx=0; rule_idx < MAX_RULES; rule_idx++) {
+        char *arg;
+        DBUS_REQ_ARG(arg, "/vm/%s/rpc-firewall-rules/%d",
+                     dom->uuid, rule_idx);
+
+        char *rulestring = db_query(conn, arg);
+
+        free(arg);
+
+        if (!rulestring)
+            break;
+
+        struct rule *policy_rule = &(dom->rules[rule_idx]);
+
+        if (create_rule(policy_rule, rulestring) < 0)
+            free_rule(*policy_rule);
+        else
+            dom->count++;
+
+        free(rulestring);
+    }
+
+}
+
+static void build_etc_policy(struct etc_policy *etc, const char *rule_filepath)
+{
+    struct stat policy_stat;
+    etc->count = 0;
+
+    if (stat(rule_filepath, &policy_stat) < 0) {
+        DBUS_BROKER_WARNING("/etc policy stat of file <%s> failed %s",
+                             rule_filepath, strerror(errno));
+        return;
+    }
+
+    char *filename = basename(rule_filepath);
+
+    if (filename[0] == '\0') {
+        DBUS_BROKER_WARNING("/etc policy invalid file <%s>", rule_filepath);
+        return;
+    }
+
+    etc->filename = filename;
+
+    size_t policy_size = policy_stat.st_size;
+
+    // There is no officially defined limit on policy file size.
+    // Rpc-broker defines a buffer size and for now limit based on that size.
+    if (policy_size > ETC_MAX_FILE) {
```

Comment:
I totally agree that `fopen()` and `getline()` would have been the more concise way to write that function.  I ended up reading the entirety of the policy file into memory for the situation, in that, if there are any changes to said file there would be a hash and/or buffer to compare and log any changes that may have been written.

There is signal handler that upon `sighup` allows the policy file to be re-built.  I was wanting to add an extra bit of monitoring in case of this event.  Perhaps there is a better way?